### PR TITLE
perf(math): add fast reduction for a special case

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR implements followings

1. Adds faster reductions for `addition`, `subtraction` and `negation`  when the modulus of the prime field is less than 2³¹ + 1.
2. Removes redundant `Reduce()` function calls
3. Enables 32-bit custom reductions for small prime fields.

### CODE REVIEW

File: tachyon/math/finite_fields/generator/prime_field_generator/build_defs.bzl [hunk 1]
```diff
@@ -35,6 +35,9 @@ def _do_generate_prime_field_impl(ctx, type):
     if ctx.attr.use_asm:
         arguments.append("--use_asm")
 
+    if ctx.attr.use_montgomery_reduce:
+        arguments.append("--use_montgomery_reduce")
```
Comment (Insun35) [inline tachyon/math/finite_fields/generator/prime_field_generator/build_defs.bzl:39]:
```suggestion
        arguments.append("--use_montgomery_reduction")
```

File: tachyon/math/finite_fields/generator/prime_field_generator/prime_field_generator.cc [hunk 2]
```diff
@@ -340,6 +345,8 @@ int RealMain(int argc, char** argv) {
       .set_required();
   parser.AddFlag<base::StringFlag>(&config.reduce).set_long_name("--reduce");
   parser.AddFlag<base::BoolFlag>(&config.use_asm).set_long_name("--use_asm");
+  parser.AddFlag<base::BoolFlag>(&config.use_montgomery_reduce)
+      .set_long_name("--use_montgomery_reduce");
```
Comment (Insun35) [inline tachyon/math/finite_fields/generator/prime_field_generator/prime_field_generator.cc:349]:
```suggestion
      .set_long_name("--use_montgomery_reduction");
```

**Review by dongchangYoo — APPROVED** [review]
LGTM

File: tachyon/math/circle/stark/BUILD.bazel [hunk 3]
```diff
@@ -31,7 +31,8 @@ generate_prime_fields(
     # Hex: 0x80000000
     modulus = "2147483648",
     namespace = "tachyon::math::stark",
-    reduce = "    return v & (static_cast<uint32_t>(kModulus[0]) - 1);",
+    reduce32 = "return v & (static_cast<uint32_t>(kModulus[0]) - 1);",
+    reduce64 = "return v & (static_cast<uint32_t>(kModulus[0]) - 1);",
```
Comment (Insun35) [inline tachyon/math/circle/stark/BUILD.bazel:35]: Isn't it better to use snake case for our naming convention?
```suggestion
    reduce_32 = "return v & (static_cast<uint32_t>(kModulus[0]) - 1);",
    reduce_64 = "return v & (static_cast<uint32_t>(kModulus[0]) - 1);",
```

Comment (chokobole) [reply tachyon/math/circle/stark/BUILD.bazel:35]: I think it's okay!

**Review by Insun35 — APPROVED** [review]
LGTM

File: tachyon/math/finite_fields/small_prime_field.h [hunk 4]
```diff
@@ -198,9 +198,9 @@ class PrimeField<_Config, std::enable_if_t<!_Config::kIsSpecialPrime &&
     EGCD<int64_t>::Result result = EGCD<int64_t>::Compute(value_, GetModulus());
     DCHECK_EQ(result.r, 1);
     if (result.s > 0) {
-      return PrimeField(Config::Reduce(result.s));
+      return PrimeField(result.s);
     } else {
-      return PrimeField(Config::Reduce(uint64_t{GetModulus()} + result.s));
+      return PrimeField(int64_t{GetModulus()} + result.s);
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/small_prime_field.h:203]: Is there any reason for changing the casting of `GetModulus()` from `uint64_t` to `int64_t`?

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

**Comment by TomTaehoonKim** [comment]
Can you fix the commit message of [c767c64](https://github.com/kroma-network/tachyon/pull/404/commits/c767c64d0f383863f2507a9e10514b2c26c2388a) from `reduce()` to `Reduce()`?

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Review by ashjeong — APPROVED** [review]
LGTM

**Pull request merged**